A few tests suggestions to see if I'm on the right track

shoffman-percona opened the pull request:
Lori Larusso mentioned you were looking for help building out the criteria for the CCC project.  We happen to have pretty deep expertise in the DB realm so I can certainly contribute more items along these lines but before I went to far I figured I'd do just a few to get feedback.  

If there's a particular area you'd like focus on, please just let me know!

eddie-knight commented:
Hey @shoffman-percona, thanks for this PR! 

Damien and Michael are approvers for this domain, but I'd like to help from a steering perspective. It occurred to me while reviewing your changes that the best place to invest time may be the threat or control definitions themselves. The current progress on the RDMS entry is iterative, and I expect it needs strong revisions before we can create accurate behaviors/steps.

Would you like to contribute to the [threat](https://github.com/finos/common-cloud-controls/blob/main/docs/governance/community-guidelines/content-standards-and-practices/threat-definitions.md) or [control definition](https://github.com/finos/common-cloud-controls/blob/main/docs/governance/community-guidelines/content-standards-and-practices/control-definitions.md) refinement?

I don't mind spending a cycle to help create runway here, if you're interested in picking it up!

shoffman-percona commented:
@eddie-knight,  Absolutely...I just started where Lori pointed me but let me comb through the threats materials.  I assume [this](https://github.com/finos/common-cloud-controls/blob/main/services/database/relational/threats.md) is where you'd want some help building out?

It may end up being a 1:x mapping of threats to controls so likely will end up coming up with them together.

eddie-knight commented:
Yes! Although the RDMS iteration is in need of an update to match the finalized guidelines (yaml).

Feel free to drop me an email (eknight@sonatype.com) or message me on the FINOS slack if I can help with implementation details.

damienjburks commented on services/database/relational/tests/ccc-rdms-c2.feature:
```diff
@@ -5,20 +5,40 @@ Feature: Secure Database Access Control
+Scenario: Ensure that authorized accounts only have the minimum neccessary permissions to perform their task
+   Given that local accounts must be granted certain permissions to perform certain operations on the database system
+   When auditing local account permissions
+   Then the permissions are the minimum needed to
```
```suggestion
   Then the permissions are the minimum needed to local account to perform necessary operations
```

damienjburks commented on services/database/relational/tests/ccc-rdms-c8.feature:
```diff
@@ -0,0 +1,18 @@
+@CCC.RDMS.C8.TR01.T02
+Scenario: Ensure ability to audit for any users who have connected using an insecure protocol
+   Given a user connection has been established to the database
+   When an admin establishes an admin connection to a database server and runs from mysql.user where ssl_type=''
```
For our controls, we try to keep them as abstract as possible. we should remove the mysql.user where ssl_type='' from this statement.

shoffman-percona replied:
Makes sense!  If I'm otherwise on track, I'll push a few more over in addition to threats/controls.

eddie-knight replied:
Also remember to keep an eye on the new [ID format](https://github.com/finos/common-cloud-controls/blob/main/docs/governance/community-guidelines/content-standards-and-practices/identifiers.md). Most notably the Ts... threats are now `TH00`, Test requirements are `TR00` and Tests are `TE00`

shoffman-percona commented:
Ok, accepted proposed changes, reworked the one that was too technology specific and also converted threats.md to threats.yaml.  hopefully this is in acceptable form (just let me know if it's not) and I'll start adding threats and converting controls.md

eddie-knight commented on services/database/relational/tests/ccc-rdms-c2.feature:
```diff
@@ -5,20 +5,41 @@ Feature: Secure Database Access Control
 """
 
-@CCC.RDMS.C2.TR02.T01
+@CCC.RDMS.C2.TR01.T01
```
```suggestion
@CCC.RDMS.C2.TR01.TE01
```

eddie-knight commented on services/database/relational/tests/ccc-rdms-c2.feature:
```diff
@@ -5,20 +5,41 @@ Feature: Secure Database Access Control
    Then the user should be granted access to the database resources
 
-@CCC.RDMS.C2.TR02.T02
+@CCC.RDMS.C2.TR01.T02
```
```suggestion
@CCC.RDMS.C2.TR01.TE02
```

eddie-knight commented on services/database/relational/tests/ccc-rdms-c2.feature:
```diff
@@ -5,20 +5,41 @@ Feature: Secure Database Access Control
    Then the user should be denied access to the database resources
 
-@CCC.RDMS.C2.TR02.TR03
+@CCC.RDMS.C2.TR01.TR03
```
```suggestion
@CCC.RDMS.C2.TR01.TE03
```

eddie-knight commented on services/database/relational/tests/ccc-rdms-c2.feature:
```diff
@@ -5,20 +5,41 @@ Feature: Secure Database Access Control
+This feature targets database configurations where a local user is defined and granted permissions to interact with the database system.  
+"""
+
+@CCC.RDMS.C2.TR02.TR01
```
```suggestion
@CCC.RDMS.C2.TR02.TE01
```

eddie-knight commented on services/database/relational/tests/ccc-rdms-c2.feature:
```diff
@@ -5,20 +5,41 @@ Feature: Secure Database Access Control
+   Then only expected local accounts exist in the database
+   And each account is properly scoped to the expected permissions
+
+@CCC.RDMS.C2.TR02.TR02
```
```suggestion
@CCC.RDMS.C2.TR02.TE02
```

eddie-knight commented on services/database/relational/tests/ccc-rdms-c2.feature:
```diff
@@ -5,20 +5,41 @@ Feature: Secure Database Access Control
+
+@CCC.RDMS.C2.TR02.TR02
+Scenario: Ensure that authorized accounts only have the minimum neccessary permissions to perform their task
+   Given that local accounts must be granted certain permissions to perform certain operations on the database system
```
```suggestion
   Given a local database with user accounts
```

Since this is all still pre-release and open to further iteration, I'm not personally concerned with getting it perfect prior to merge... but adding some notes anyhow.

We should keep in mind the `Given` step is intended to represent a base state or precondition that can be checked prior to the tests being run.

In this example, I'm imagining that we want to use this step to report whether there are non-admin accounts present _before_ going on to report whether all non-admin accounts are correctly provisioned.

eddie-knight commented on services/database/relational/tests/ccc-rdms-c8.feature:
```diff
@@ -0,0 +1,18 @@
+This reature ensures that end-to-end encryption of data in transit is leveraged and enforced
+"""
+
+@CCC.RDMS.C8.TR01.T01
```
```suggestion
@CCC.RDMS.C8.TR01.TE01
```

eddie-knight commented on services/database/relational/tests/ccc-rdms-c8.feature:
```diff
@@ -0,0 +1,18 @@
+   When the connection attempt is made without using encryption
+   Then the connection should be refused
+
+@CCC.RDMS.C8.TR01.T02
```
```suggestion
@CCC.RDMS.C8.TR01.TE02
```

eddie-knight commented on services/database/relational/threats.yaml:
```diff
@@ -0,0 +1,49 @@
+title: Relational Database Management Systems Threats
```
I think this file may be following an older schema. The latest guidance on this topic can be found in the project governance directory.

ref:
- https://github.com/finos/common-cloud-controls/blob/main/docs/governance/community-guidelines/content-standards-and-practices/threat-definitions.md 
- https://github.com/finos/common-cloud-controls/blob/main/docs/governance/community-guidelines/templates/threats.yaml

shoffman-percona replied:
Updated to the new schema although I left description at the highest level as it made sense to keep to me.  No worries if it gets dropped.

eddie-knight reviewed (commented):
Added some comments that I hope help accelerate your progress.

Happy to help contribute to the gherkin writing, perhaps in something like google docs so that we can have faster iterations— these things are hard to get right!

shoffman-percona commented:
Google would be easier to co-edit but staying in the repo is helping me learn where all the relevant docs are located.  I was guessing a bit earlier but now and seeing how to find definitions and templates and such.
